staging/src/k8s.io/client-go/rest/with_retry.go

```diff
@@ -372,6 +379,63 @@ func readAndCloseResponseBody(resp *http.Response) {
 	}
 }
 
+type ResponseCloser struct {
+	once     sync.Once
+	resp     *http.Response
+	respCopy io.ReadWriteCloser
+	err      error
+}
+
+func NewResponseCloser(resp *http.Response) *ResponseCloser {
+	return &ResponseCloser{
+		resp:     resp,
+		respCopy: &buffer{b: &bytes.Buffer{}},
+	}
+}
+
+func (r *ResponseCloser) DrainAndClose() error {
```

Comment:
since we only need to do this buffering when called from inside prepareNextRetry, can take a param to make the buffering optional, and pass `false` when we call this everywhere except prepareNextRetry